api/saml/wayfless.py

```diff
@@ -0,0 +1,179 @@
+    ) -> SAMLWAYFlessConfiguration:
+        """Return the WAYFless configuration object.
+
+        :param db: SQLAlchemy session
+        :return: SAMLWAYFlessConfiguration object
+        """
+        with self._configuration_factory.create(
+            self._configuration_storage, db, SAMLWAYFlessConfiguration
+        ) as configuration:
+            yield configuration
+
+    def _get_wayfless_url_template(
+        self, db: sqlalchemy.orm.session.Session
+    ) -> Optional[str]:
+        """Return a templated acquisition link.
+
+        :param db: SQLAlchemy session
+        :return: Templated acquisition link
+        """
+        with self._get_configuration(db) as configuration:
+            return configuration.wayfless_url_template
+
+    def external_integration(
+        self, db: sqlalchemy.orm.session.Session
+    ) -> ExternalIntegration:
+        """Return an ExternalIntegration object associated with the collection with a WAYFless url.
+
+        :param db: SQLAlchemy session
+        :return: ExternalIntegration object
+        """
+        return get_one(db, ExternalIntegration, id=self._external_integration_id)
+
+    def fulfil(
+        self, patron, pin, licensepool, delivery_mechanism, fulfillment: FulfillmentInfo
+    ) -> FulfillmentInfo:
+        db = Session.object_session(patron)
+        acquisition_link_template = self._get_wayfless_url_template(db)
+
+        self._logger.debug(
+            f"WAYFless acquisition link template: {acquisition_link_template}"
+        )
+
+        if acquisition_link_template:
+            saml_credential = self._saml_credential_manager.lookup_saml_token_by_patron(
+                db, patron
+            )
+
+            self._logger.debug(f"SAML credentials: {saml_credential}")
+
+            if saml_credential:
+                saml_subject = self._saml_credential_manager.extract_saml_token(
+                    saml_credential
+                )
+
+                self._logger.debug(f"SAML subject: {saml_subject}")
+
+                if not saml_subject.idp:
+                    raise SAMLWAYFlessFulfilmentError(
+                        f"SAML subject {saml_subject} does not contain an IdP's entityID"
+                    )
+
+                acquisition_link = acquisition_link_template.replace(
+                    SAMLWAYFlessConfiguration.IDP_PLACEHOLDER,
+                    urllib.parse.quote(saml_subject.idp, safe=""),
+                )
+                acquisition_link = acquisition_link.replace(
+                    SAMLWAYFlessConfiguration.ACQUISITION_LINK_PLACEHOLDER,
+                    urllib.parse.quote(fulfillment.content_link, safe=""),
+                )
+
+                self._logger.debug(
+                    f"Old acquisition link {fulfillment.content_link} has been transformed to {acquisition_link}"
+                )
+
+                fulfillment.content_link = acquisition_link
+
+                return fulfillment
```

Comment:
There are the following cases:
1. No `acquisition_link_template`. In this case CM must return the existing `FulfillmentInfo` object.
2. `acquisition_link_template` exists but there is no `saml_credential`. In this case I think we need to raise an error: we expect the patron to be authenticated using SAML but there is no `Credential` object in the database. It probably means that there is some configuration error and either SAML Auth Provider or collection's `acquisition_link_template` are not configured properly. Returning the existing `FulfillmentInfo` object will not work because content provider expects a WAYFless URL.
3. `acquisition_link_template`, `saml_credential` are present but there is no `saml_subject.idp`. In this case we need to raise an error because we cannot form the WAYFless URL without IdP's entityID. Returning the existing `FulfillmentInfo` object will again not work because the content provider expects a WAYFless URL.